File: imd/browser.js
ObjC.import("Foundation");

function endpoint(url) {
  const parts = $.NSURLComponents.componentsWithString($(url));
  if (!parts) return null;
  const scheme = ObjC.unwrap(parts.scheme)?.toLowerCase();
  const host = ObjC.unwrap(parts.host)?.toLowerCase();
  if (!host || !["http", "https"].includes(scheme)) return null;
  const port = ObjC.unwrap(parts.port) ?? (scheme === "https" ? 443 : 80);
  return JSON.stringify([scheme, host, port]);
}

function openChrome(url) {
  const wanted = endpoint(url);
  if (!wanted) throw new Error("Use a complete HTTP or HTTPS URL.");
  const chrome = Application("com.google.Chrome");
  let windows = chrome.windows();
  for (const win of windows) {
    const tabs = win.tabs();
    for (let index = 0; index < tabs.length; index++) {
      if (endpoint(tabs[index].url()) !== wanted) continue;
      win.activeTabIndex = index + 1;
      win.minimized = false;
      win.index = 1;
      chrome.activate();
      return { action: "focused" };
    }
  }
  if (!windows.length) {
    chrome.windows.push(chrome.Window());
    windows = chrome.windows();
    windows[0].tabs()[0].url = url;
    windows[0].activeTabIndex = 1;
  } else {
    const win = windows[0];
    win.tabs.push(chrome.Tab({ url }));
    win.activeTabIndex = win.tabs().length;
  }
  windows[0].minimized = false;
  windows[0].index = 1;
  chrome.activate();
  return { action: "opened" };
}

function run(args) {
  return JSON.stringify(openChrome(args[0]));
}

File: imd/browser.py
"""Open an HTTP URL in Chrome or select an existing tab."""

import json
import subprocess
import sys
from pathlib import Path
from threading import Lock
from urllib.parse import urlsplit

_lock = Lock()
_script = Path(__file__).with_name("browser.js")


def open_url(url: str) -> dict[str, str]:
    """Select a tab with the same protocol, host, and port, or open a tab."""
    try:
        parts = urlsplit(url)
        valid = (
            parts.scheme in {"http", "https"}
            and parts.hostname
            and (parts.port is None or 1 <= parts.port <= 65535)
            and not any(ord(char) < 33 for char in url)
            and "\\" not in url
        )
    except ValueError:
        valid = False
    if not valid:
        raise ValueError("Use a complete HTTP or HTTPS URL with a valid port.")
    if sys.platform != "darwin":
        raise RuntimeError("Cmd + click requires macOS and Google Chrome.")
    with _lock:
        try:
            result = subprocess.run(
                ["osascript", "-l", "JavaScript", str(_script), url],
                capture_output=True,
                text=True,
                timeout=30,
                check=False,
            )
        except subprocess.TimeoutExpired as exc:
            raise RuntimeError(
                "Chrome does not respond. Check the macOS Automation permission and try again."
            ) from exc
        except OSError as exc:
            raise RuntimeError("macOS browser automation is not available.") from exc
    if result.returncode:
        raise RuntimeError(
            "Cannot control Google Chrome. Install Chrome and allow the app that starts IMD "
            "to control Chrome in System Settings > Privacy & Security > Automation."
        )
    try:
        data = json.loads(result.stdout)
        if data.get("action") not in {"focused", "opened"}:
            raise ValueError
    except (ValueError, AttributeError) as exc:
        raise RuntimeError("Chrome returns an invalid response.") from exc
    return {"action": data["action"]}
